### PR TITLE
Temp list variable documentation

### PR DESCRIPTION
**Pull request by sanderr**
# Description

- Made an attempt to clear up confusing nature of `TempListVariable` naming and semantics. @wouterdb could you check for correctness? @arnaudsjs could you check for clarity?
- Introduced ABC for result variables associated with a relation attribute. @arnaudsjs I added this because I believed it might help with determining which `DelayedResultVariable` instances are relevant for #3098, because your remark yesterday got me thinking that this wasn't very clear in the current naming and typing. If you don't think this adds value I can revert this part.

loosely related to #3098

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [ ] Attached issue to pull request
- [ ] Changelog entry
- [ ] Type annotations are present
- [ ] Code is clear and sufficiently documented
- [ ] No (preventable) type errors (check using make mypy or make mypy-diff)
- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

# Reviewer Checklist:

- [ ] Sufficient test cases (reproduces the bug/tests the requested feature)
- [ ] Code is clear and sufficiently documented
- [ ] Correct, in line with design

### CODE REVIEW

File: src/inmanta/execute/runtime.py [hunk 1]
```diff
@@ -173,18 +173,29 @@ def get_dataflow_node(self) -> dataflow.AssignableNodeReference:
         return self._node
 
 
-class AttributeVariable(ResultVariable["Instance"]):
+class RelationAttributeVariable(ABC):
```
Comment (wouterdb) [inline src/inmanta/execute/runtime.py:176]: Doesn't extending ABC have a significant performance overhead?  (For no reason that could be achieved by documentation alone),?


(these objects are constructed by the thousands)

File: src/inmanta/execute/runtime.py [hunk 2]
```diff
@@ -392,27 +403,37 @@ def __str__(self) -> str:
         return "BaseListVariable %s" % (self.value)
 
 
-class TempListVariable(BaseListVariable):
+class ListLiteral(BaseListVariable):
+    """
+    Transient variable to represent a list (of either constants or instances) literal (not a variable).
```
Comment (wouterdb) [inline src/inmanta/execute/runtime.py:408]: The size of the list is fixed/known, which causes the promises tracking to be fully accurate (freezing is about guessing the size,...)

**Review by arnaudsjs — APPROVED** [review]

File: src/inmanta/execute/runtime.py (hunk 2 above)
Comment (sanderr) [reply src/inmanta/execute/runtime.py:408]: Could you elaborate? I agree with what you're saying but I don't understand what you mean by the comment. As far as I can see it's just another way of saying what the docstring says.

File: src/inmanta/execute/runtime.py (hunk 1 above)
Comment (sanderr) [reply src/inmanta/execute/runtime.py:176]: I don't immediately see how, but you're right that there's no strong reason to do so, I'll get rid of it.

**Comment by sanderr** [comment]
@arnaudsjs could you pay special attention to [a7ffe55](https://github.com/inmanta/inmanta-core/pull/3842/commits/a7ffe552fb2bda9960124618ab6e057e2c3b187d)?

File: src/inmanta/execute/scheduler.py [hunk 3]
```diff
@@ -525,16 +525,11 @@ class PrioritisedDelayedResultVariableQueue:
...
-      constraints.
-    * Finally, all TempListVariables are returned.
-
-    TempListVariables is subclass of DelayedResultVariables that is not
-    associated with an entity. That way it's not possible to set
-    a relation precedence rule on them.
+    * First return the DelayedResultVariables for relations that do not have an order constraint.
+    * Then return DelayedResultVariables for relations with order constraint.
+      They are returned in an order that is valid with respect to the constraints.
+    * Finally, all variables DelayedResultVariables that are not associated with an entity are
```
Comment (arnaudsjs) [inline src/inmanta/execute/scheduler.py:531]:
```suggestion
    * Finally, all DelayedResultVariables that are not associated with an entity are
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

**Review by wouterdb — APPROVED** [review]

**Pull request closed without merging**